torch/fx/experimental/proxy_tensor.py

```diff
@@ -980,12 +985,25 @@ def __init__(self, base, path):
 
             def __getattr__(self, name):
                 assert isinstance(self, torch.nn.Module)
+                # Calling into torch.nn.Module.__getattr__ with super(),
+                # That __getattr__ is patched to be module_getattr_wrapper in _symbolic_trace.py.
+                # which then calls into _ModuleStackTracer.getattr
                 attr_val = super().__getattr__(name)
                 if isinstance(attr_val, AttrProxy):
                     attr_val = self_.proxy_modules[attr_val]
                 elif not isinstance(attr_val, torch.nn.Module):
                     return attr_val
-                return AttrProxy(attr_val, self_.proxy_paths[self] + "." + name)
+                if attr_val not in self_.attr_proxy_map:
+                    self_.attr_proxy_map[attr_val] = AttrProxy(attr_val, self_.proxy_paths[self] + "." + name)
+                else:
+                    # NOTE [caching AttrProxy]. Caching ensures a 1-1 mapping between AttrProxy and the actual attr_val.
+                    # 1. This is to solve the diamond shape reference problem: we want to record the path as A.B.D
+                    # instead of A.C.D.
+                    # 2. Instead of creating a new AttrProxy, we just reset the path of existing one. This is to avoid
+                    # dynamo creating multiple guards for the same attr_val but different AttrProxy when exporting
+                    # a model that calls torch.compile (e.g when a model uses torch.cond.)
+                    self_.attr_proxy_map[attr_val].reset(attr_val, self_.proxy_paths[self] + "." + name)
+                return self_.attr_proxy_map[attr_val]
```

Comment:
I am not familiar with this part of the code. Maybe @tugsbayasgalan is the right person?